sort_image_files.sh

```diff
@@ -0,0 +1,223 @@
+#
+# Note: Photo-taken date is exif:DateTimeOriginal. If no exif:DateTimeOriginal, 
+# script uses date:modify.
+#
+# Author: Kim Lew
+
+if [ $# -gt 3 ]; then 
+  # Case of 4 or more parameters given.
+  echo "Give 3 command-line arguments. Or give 0 arguments & get prompts."
+  exit 1
+fi
+
+if [ $# -eq 0 ]; then
+  echo "Type the directory path for the files you want to sort: "
+  read -r directory_path
+  echo "Do you want Day sub-directories, along with the Year & Month ones? [y/n] "
+  read -r day_subdir_also
+  echo "Do you want to re-name the files with IMG, replacing IMG with the date? [y/n] "
+  read -r rename_files_also
+elif [[ $# -eq 1 || $# -eq 2 || $# -eq 3 ]]; then
+  if [ $# -eq 1 ]; then
+    echo "For parameter 2: Type y if you also want Day subdirectories."
+    echo "For parameter 3: Type y if you also want to re-name the files with IMG using the date."
+    exit 1
+  elif [ $# -eq 2 ]; then
+    echo "For parameter 3: Type y if you also want to re-name the files with IMG using the date."
+    exit 1
+  elif [ $# -eq 3 ]; then
+    directory_path="$1"
+    day_subdir_also="$2"
+    rename_files_also="$3"
+  fi
+fi
+
+if [ ! -d "$directory_path" ]; then
+    echo "This directory does NOT exist." 
+    exit 1
+fi
+
+if ! magick identify --version > /dev/null; then
+  echo "Error: You are missing the identify program that is part of the"
+  echo "ImageMagick software suite. Install ImageMagick with your package manager."
+  echo "Or see: https://imagemagick.org/index.php/"
+  echo "Then re-run this script."
+  exit 1
+fi
+
+echo "Path you gave:" "${directory_path}"
+echo "Day subdirectories wanted?: " "${day_subdir_also}"
+echo "Rename files?: " "${rename_files_also}"
+
+case "${day_subdir_also}" in
+  [yY] | [yY][eE][sS]) # To create year-month-day subdirectories.
+    clean_day_subdir_also='y'
+;;
+  [nN] | [nN][oO] | '') # To create ONLY year-month subdirectories.
+    clean_day_subdir_also='n'
+;;
+    *)
+echo "Invalid input for 2nd parameter given or answer to 2nd prompt. Type y or n."
+exit 1
+;;
+esac
+case "${rename_files_also}" in
+  [yY] | [yY][eE][sS]) # To create year-month-day subdirectories.
+    clean_rename_files_also='y'
+;;
+  [nN] | [nN][oO] | '') # To create ONLY year-month subdirectories.
+    clean_rename_files_also='n'
+;;
+    *)
+echo "Invalid input for 3rd parameter given or answer to 3rd prompt. Type y or n."
+exit 1
+;;
+esac
+ 
+echo
+echo "SORTING files..."
+
+file_sort_counter=0
+
+# Loop that processes entire given directory.
+while read -r a_file_name; do
+  echo "Looking at file:" "${a_file_name}"
+  exif_date="$(identify -format '%[EXIF:DateTimeOriginal]' "$a_file_name" 2> /dev/null)" 
+  modify_date="$(identify -format '%[DATE:modify]' "$a_file_name" 2> /dev/null)"
+  echo -e " Exif Date is: \t\t" "${exif_date}"
+  echo -e " Modify Date is: \t" "${modify_date}"
+  
+  if [[ "$exif_date" == '' ]] && [[ -z "${modify_date}" ]]; then
```

Comment:
Note that the tests `[ -z "${SOME_VAR}" ]` and `[ "${SOME_VAR}" == '' ]` are equivalent, so you could use a consistent style like

```shell
if [[ -z "${exif_date}" ]] && [[ -z "${modify_date}" ]]; then
```

Also, since you're using the `[[ ]]` form for the testing command, you can use a `&&` operator _inside_ the double square brackets, like this:

```shell
if [[ -z "${exif_date}" && -z "${modify_date}" ]]; then
```

Reply:
@kcharter That -z isn't very intuitive so I went with the '' empty string version. Done.

Reply:
@kimlew whatever you find easiest to read -- looks good.